content/en/blog/_posts/2024-11-10-migrate-cgroup-v2.md

```diff
@@ -0,0 +1,162 @@
+2. PSI is planned in a future release [KEP-4205](https://github.com/kubernetes/enhancements/issues/4205),
+   but pending due to runc 1.2.0 release delay.
+3. monitoring tools support, like [Cadvisor](https://github.com/google/cadvisor/).
+   Currently, cgroups v2 features are not fully-supported yet.
+
+## Adopting cgroup version 2
+
+### Requirements
+
+Here's what you need to use cgroup v2 with Kubernetes.
+First up, you need to be using a version of Kubernetes with support for v2 cgroup management;
+that's been stable since Kubernetes v1.25 and all supported Kubernetes releases include this support.
+
+- OS distribution enables cgroups v2
+- Linux Kernel version is 5.8 or later
+- Container runtime supports cgroups v2. For example:
+  - containerd v1.4 or later (at the time of writing, containerd releases v1.6 and later are within that project's support period)
+  - CRI-O v1.20 or later
+- The kubelet and the container runtime are configured to use the systemd cgroup driver
+
+#### kernel updates around cgroups v2
+
+cgroups v2 first appeared in Linux Kernel 4.5 in 2016.
+
+- In Linux 4.5, cgroups v2 `io`, `memory` & `pid` cgroups management were supported.
+- Linux 4.15 added support for cgroups v2 `cpu` management
+- [Pressure Stall Information](https://docs.kernel.org/accounting/psi.html) (PSI) support began with Linux 4.20.
+- The Kubernetes project does not recommend using cgroups v2 with a Linux kernel older than 5.2 due to lack of cgroup-level task freezer support.
+- In Kubernetes, 5.8 is the minimal kernel version for cgroups v2 as root `cpu.stat` file on cgroupv2
+  was only added on kernel 5.8.
+- `memory.peak` is added in 5.19.
+
+### Use systemd as cgroup driver
+
+[Configure the kubelet's cgroup driver to match the container runtime cgroup driver](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver/).
+
+The [Container runtimes](/docs/setup/production-environment/container-runtimes)
+page explains that the `systemd` driver is recommended for kubeadm
+based setups instead of the kubelet's default `cgroupfs` driver,
+because kubeadm manages the kubelet as a systemd service.
+
+A minimal example of configuring the field explicitly:
+
+```yaml
+apiVersion: kubelet.config.k8s.io/v1beta1
+kind: KubeletConfiguration
+cgroupDriver: systemd
+```
+
+In v1.31, [KEP-4033](https://github.com/kubernetes/enhancements/issues/4033) is beta to extend CRI API for the kubelet
+to discover the cgroup driver from the container runtime. This will help installer and kubelet to autodetect
+
+- TODO
+
+### Tools and commands for troubleshooting
+
+Tools and commands that you should know about cgroups:
+
+- `stat -fc %T /sys/fs/cgroup/`: Check if cgroups v2 is enabled which will return `cgroup2fs`
+- `systemctl list-units kube* --type=slice` or `--type=scope`: List kube related units that systemd currently has in memory.
+- `bpftool cgroup list /sys/fs/cgroup/*`: List all programs attached to the cgroup CGROUP.
```

Comment:
`bpftool cgroup list /sys/fs/cgroup/*` actually shows all programs?
In my environment, I could not list it.
First of all, I think `bpftool cgroup tree` is suitable to list all  attached  programs.

BTW, I have recently fixed the bug of `bpftool cgroup` as below:
https://patchwork.kernel.org/project/netdevbpf/patch/20240607111704.6716-1-tadakentaso@gmail.com/
So if you want to try bpftool cgroup tree` in your environment, please keep that in your mind.